core/ordersync/ordersync.go

```diff
@@ -285,50 +289,79 @@ func (s *Service) GetOrders(ctx context.Context, minPeers int) error {
 		default:
 		}
 
-		// TODO(albrow): As a performance optimization, do this for loop
-		// partly in parallel.
+		m := &sync.Mutex{}
+		wg := &sync.WaitGroup{}
+		semaphore := make(chan struct{}, maxRequestPeersInParallel)
 		currentNeighbors := s.node.Neighbors()
 		shufflePeers(currentNeighbors)
+		innerCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
 		for _, peerID := range currentNeighbors {
-			if len(successfullySyncedPeers) >= minPeers {
+			m.Lock()
+			successfullySyncedPeerLength := len(successfullySyncedPeers)
+			successfullySynced := successfullySyncedPeers.Contains(peerID.Pretty())
```

Comment:
@albrow This shouldn't need to be in the mutex unless we have duplicate peer IDs in the `currentNeighbors` array. My guess is that we don't, but I also wanted to protect us from bugs in libp2p that could cause this assumption to not be true. Let me know if you think it's okay to access this outside of a lock.

Reply:
It doesn't matter whether there are any duplicate peer IDs. [Maps in Go are not goroutine safe](https://golang.org/doc/faq#atomic_maps). We need to protect reads and writes to the map with a `sync.RWMutex`.

Reply:
Oh I see. I didn't realize that it could crash the program -- I assumed that it was merely a synchronization issue. I'll leave it in the mutex and change that to a `RWMutex`.